Cabal/Distribution/Types/Version.hs

```diff
@@ -96,8 +97,25 @@ instance Pretty Version where
                                 (map Disp.int $ versionNumbers ver))
 
 instance Parsec Version where
-    parsec = mkVersion <$> P.sepBy1 P.integral (P.char '.') <* tags
+    parsec = do
+        digit <- digitParser <$> askCabalSpecVersion
+        mkVersion <$> P.sepBy1 digit (P.char '.') <* tags
       where
+        digitParser v
+            | v >= CabalSpecV1_24 = P.integral
+            | otherwise           = (some d >>= toNumber) P.<?> "non-leading-zero integral"
+          where
+            toNumber :: CabalParsing m => [Int] -> m Int
+            toNumber [0] = return 0
+            toNumber xs@(0:_) = do
+                parsecWarning PWTVersionLeadingZeros "Version digit with leading zero. Use cabal-version: 1.24 or later to write such versions. For more information see https://github.com/haskell/cabal/issues/5092"
```

Comment:
It's a bit more tricky; while technically cabal 1.24 happily copes with this, cabal 1.22 would not, as cabal 1.22 would still try to parse future spec versions, and then choke on unexpected syntax.

So as far as `cabal check`'s category of "inacceptable things for distribution via hackage" errors is concerned, we should probably require at least `cabal-version:2.0`; *or* we may consider leading zeroes a mistake, and prevent them from being for *any* .cabal file uploaded to hackage.

Reply:
I'll change the cutoff to 2.0.

If we want to completely prevent them from Hackage is something I don't feel confident to decide right now (even it will affect only few packages)